### PR TITLE
output/cloudv2: Binary-based payload

### PR DESCRIPTION
**Pull request by codebien**
It adds the HTTP client for flushing the metric to the remote cloud service and the new Protobuf-based required payload.

Updates #2954

### CODE REVIEW

File: cloudapi/config.go
Comment (na--) [inline cloudapi/config.go:31]: Maybe this can be something a bit more flexible?

```suggestion
	UsedAPIVersion    null.Int   `json:"usedVersion" envconfig:"K6_CLOUD_USED_VERSION"`
```

And, if the used version is higher than the supported one, k6 can exit with an error? :thinking:

File: output/cloud/output.go
Comment (na--) [inline output/cloud/output.go:49]:
```suggestion
	SomeCloudOutputSubInterface
```

Instead of this embedded `*expv2.Output` and all of the `if !out.config.UseVersion2.Bool` and `if out.outv2` statements sprinkled in all of the methods, it seems to me that it would be easier to add a new small `interface` type here with a few methods. Potentially a sub-set of the `output.Output` interface.

Then embed that type in the struct and have 2 implementations of that interface - one for the old json-based cloud output and one for the new protobuf one. `newOutput()` can instantiate the correct one and save it in the struct instance it returns.

File: cloudapi/config.go
Comment (codebien) [reply cloudapi/config.go]: Addressed in https://github.com/grafana/k6/pull/2963/commits/ea2c1ef47959ab3ae81474440aa89009a34f86ae

**Comment by codebien** [comment]
@na-- @olegbespalov I added the aggregation proposal also, it shouldn't be optimal but it should be a good first step to achieve a reasonable rate. We can iterate on it after we will get some benchmarks (I don't plan them for this PR).

I would get the reviews for it before I do the refactor asked in https://github.com/grafana/k6/pull/2963#discussion_r1157118508 so I can address it all at once.

File: output/cloud/expv2/output.go
Comment (olegbespalov) [inline output/cloud/expv2/output.go:113]: Let's be more precise here (instead of prefixing some of the errors inside the `Push` method with `failed to push metrics:`). 

That way we avoid any bare/unwrapped error from inside.

**Review by olegbespalov — COMMENTED** [review]
I've done a first round, will do the other one a bit later

File: output/cloud/expv2/metrics_client_test.go
Comment (mstoykov) [reply output/cloud/expv2/metrics_client_test.go:42]: [ioutil.ReadAll](https://pkg.go.dev/io/ioutil#ReadAll) is the deprecated one ;)

Comment (olegbespalov) [reply output/cloud/expv2/metrics_client_test.go:42]: My bad :facepalm:

File: output/cloud/expv2/pbcloud/metric.proto
Comment (codebien) [reply output/cloud/expv2/pbcloud/metric.proto:11]: You can have multiple enums without the prefix I guess it would conflict.

Comment (codebien) [reply output/cloud/expv2/pbcloud/metric.proto:11]: The error here is not using the prefix also for the other items that is a suggested style guide https://protobuf.dev/programming-guides/style/#enums

File: output/cloud/expv2/output.go
Comment (codebien) [reply output/cloud/expv2/output.go:77]: Because the reference ID is generated when we _Start_ the output and not when we instantiate the output. The reference id is generated from the cloud backend.

File: output/cloud/expv2/pbcloud/metric.proto
Comment (codebien) [reply output/cloud/expv2/pbcloud/metric.proto]: [019b27a](https://github.com/grafana/k6/pull/2963/commits/019b27a29df6c7ef8b8620ac37d729b4f9e0bd31)

File: output/cloud/expv2/metrics_client_test.go
Comment (olegbespalov) [reply output/cloud/expv2/metrics_client_test.go:41]: I still can see that line 41 duplicates line 40

```
assert.Equal(t, "2.0", r.Header.Get("K6-Metrics-Protocol-Version"))
```

File: output/cloud/output.go
Comment (olegbespalov) [inline output/cloud/output.go:221]: I know that is obvious, and we probably don't have many versions, but I suggest avoiding using the "magic" number 2 and creating a constant. 

That way, we keep code more maintainable in the future since any other maintainer/contributor will easily find all such conditional places and won't spend (or miss) the places where we compare with some number.

File: output/cloud/expv2/output.go
Comment (olegbespalov) [inline output/cloud/expv2/output.go:81]: Probably it's an old comment :thinking:

File: output/cloud/output.go
Comment (codebien) [reply output/cloud/output.go]: @na-- I made an attempt in the last commits, they are just the final result of the process that can be fine in the history here: https://github.com/grafana/k6/tree/cloudv2

I've created a dedicated PR for the versioning gateway so it should be clear and easier to get the architecture and merge with a different timing https://github.com/grafana/k6/pull/3041

File: output/cloud/expv2/output.go
Comment (mstoykov) [inline output/cloud/expv2/output.go:267]: nit:
```suggestion
		pb.Labels = make([]*pbcloud.Label, 0, ((*atlas.Node)(ts.Tags)).Len()+2)
```

File: output/cloud/output.go
Comment (codebien) [reply output/cloud/output.go]: This is resolved now on the [root PR](https://github.com/grafana/k6/pull/3072) in the https://github.com/grafana/k6/pull/3072/commits/448da4c3217ccff1b21c07462724e13e628bc075 commit

File: output/cloud/expv2/metrics_client_test.go
Comment (codebien) [reply output/cloud/expv2/metrics_client_test.go]: The last push resolved it

File: output/cloud/expv2/metrics_client.go
Comment (olegbespalov) [inline output/cloud/expv2/metrics_client.go:102]: Is there any convention in BE, on how they structure error responses? I mean could we maybe try to extract something from there?

File: output/cloud/expv2/metrics_client.go
Comment (olegbespalov) [inline output/cloud/expv2/metrics_client.go:109]: Let's maybe name this more precisely, like `cloudProcessingLimit` ?

Comment (codebien) [reply output/cloud/expv2/metrics_client.go]: https://github.com/grafana/k6/pull/2963/commits/3d401fe96daca1144d06ff0ecb89747a5c914e68

File: output/cloud/expv2/metrics_client.go
Comment (olegbespalov) [reply output/cloud/expv2/metrics_client.go:102]: We can do that in a separated PR

**Review by olegbespalov — DISMISSED** [review]

File: output/cloud/expv2/metrics_client.go
Comment (olegbespalov) [inline output/cloud/expv2/metrics_client.go:51]: Isn't the `consts.Version` already has a `v` prefix? :thinking:

Comment (na--) [reply output/cloud/expv2/metrics_client.go:51]: Nope: https://github.com/grafana/k6/blob/4e8347ee4b12c65069e55e7c7fe82ec9aa57d777/lib/consts/consts.go#L11

**Review by na-- — DISMISSED** [review]

File: output/cloud/expv2/metrics_client.go
Comment (olegbespalov) [reply output/cloud/expv2/metrics_client.go:51]: Right :facepalm: :+1:

File: output/cloud/expv2/metrics_client.go
Comment (mstoykov) [inline output/cloud/expv2/metrics_client.go:127]: Here we have:
1. []byte that is the marshaled data
2. second []byte that is the snappied data

I would expect both won't be *that* big, but if possible I would prefer if we can user writers and readers and preferably have only 1 final copy of the `[]byte` or even the writing to happen in stream fashion.

This *does* depend on whether the backend support streaming snappy format as there are two different formats.

**Review by olegbespalov — APPROVED** [review]

File: output/cloud/expv2/metrics_client.go
Comment (mstoykov) [inline output/cloud/expv2/metrics_client.go:111]: Shouldn't we just break them in parts then? 100kb doesn't soudn *that* big. Especially before compression.

File: output/cloud/expv2/metrics_client.go
Comment (mstoykov) [inline output/cloud/expv2/metrics_client.go:79]: Unless we are writing directly to the `buf` this is pointless and actually harmful. 

You already created the `[]byte` slice and now you are going to put it in a pool so it will survive longer. Only to get it out and not use it but just write on top of it.

If we are not going to be writing to it (seem my comments on `newRequestBody`) it is better to just use `bytes.Reader` and skip the pool.

File: output/cloud/expv2/metrics_client.go
Comment (mstoykov) [inline output/cloud/expv2/metrics_client.go:37]: Does this need to be public?

File: go.mod
Comment (codebien) [reply go.mod:15]: I didn't know about it, going to change it

File: output/cloud/expv2/metrics_client.go
Comment (codebien) [reply output/cloud/expv2/metrics_client.go:111]: How would you do that? Here the limit is related to the protobuf, how do you chunk it without breaking the metric set message? I'm not aware of Protobuf supporting a streaming/chunked version. Did I miss it?

This should be an extreme check, if we need to chunk, then I think we should do that before calling the metric's client, so in its caller. WDYT?

File: go.mod
Comment (codebien) [inline go.mod]: Now, we should do the same with the Prometheus remote write so we can drop it.

File: go.mod
Comment (codebien) [reply go.mod]: Done

File: output/cloud/expv2/metrics_client.go
Comment (codebien) [reply output/cloud/expv2/metrics_client.go]: I added a TODO for doing and testing in a specific PR so we don't block this.

File: output/cloud/expv2/metrics_client.go
Comment (codebien) [reply output/cloud/expv2/metrics_client.go]: You're right, this is some remaining stuff from the old client. Done.

File: output/cloud/expv2/metrics_client.go
Comment (codebien) [reply output/cloud/expv2/metrics_client.go]: Done

File: go.mod
Comment (mstoykov) [reply go.mod]: Just make certain to test it as it might turn out there are incompatibilities ... although I hope not :crossed_fingers:

Comment (codebien) [reply go.mod]: Added a issue there

File: output/cloud/expv2/metrics_client.go
Comment (mstoykov) [reply output/cloud/expv2/metrics_client.go:103]: You can use https://pkg.go.dev/google.golang.org/protobuf/proto#Size to find out how big it will be *before* serializing and then splitting if necessary.

And I guess a method like Push can know that protobuf will be used and cut it in pieces when necessary.

We can probably leave that for after we merge everything. But I would like to know how many metrics are 100kb before we actually make this the default. As dropping your metrics because they were 101kb sounds like terrible experience.

**Review by mstoykov — APPROVED** [review]
LGTM! But we should test how big is 100kb encoded metrics before we actually make this the default way to push metrics to the cloud.

**Comment by codebien** [comment]
>But we should test how big is 100kb encoded metrics

Yes, I already have it on my list. I add it also to the worklog in https://github.com/grafana/k6/issues/2954 as a reminder.

**Review by olegbespalov — APPROVED** [review]

**Pull request merged**
